odoo/orm/models.py

```diff
@@ -197,31 +197,53 @@ class MetaModel(api.Meta):
         Its main purpose is to register the models per module.
     """
     module_to_models = defaultdict(list)
+    models = set()
 
     def __new__(meta, name, bases, attrs):
         # this prevents assignment of non-fields on recordsets
         attrs.setdefault('__slots__', ())
         # this collects the fields defined on the class (via Field.__set_name__())
         attrs.setdefault('_field_definitions', [])
 
-        if attrs.get('_register', True):
-            # determine '_module'
-            if '_module' not in attrs:
-                module = attrs['__module__']
-                assert module.startswith('odoo.addons.'), \
-                    f"Invalid import of {module}.{name}, it should start with 'odoo.addons'."
-                attrs['_module'] = module.split('.')[2]
+        if any(not isinstance(base, MetaModel) for base in bases):
+            raise TypeError('The Odoo models should only contain Odoo model without any other python classes.')
 
-            _inherit = attrs.get('_inherit')
-            if _inherit and isinstance(_inherit, str):
-                # TODO: add an exception: TypeError(f"'_inherit' property of model {name!r} should be a list: {_inherit!r}.")
-                attrs.setdefault('_name', _inherit)
-                attrs['_inherit'] = [_inherit]
-
-            if not attrs.get('_name'):
-                attrs['_name'] = class_name_to_model_name(name)
-
-        return super().__new__(meta, name, bases, attrs)
+        if not attrs.get('_register', True):
+            return super().__new__(meta, name, bases, attrs)
+
+        inherit_models = list({base._name for base in bases if base._name})
+        _inherit = attrs.get('_inherit')
+        if _inherit:
+            if bool(inherit_models) and attrs.get('_inherit'):
+                raise TypeError('Please use only the pythonic inheritance and remove "_inherit" attribute.')
+            inherit_models.extend([_inherit] if isinstance(_inherit, str) else _inherit)
+
+        if '_name' in attrs and attrs['_name'] is None and not inherit_models:
+            # BaseModel, AbstractModel or Model or TransientModel
+            return super().__new__(meta, name, bases, attrs)
+
+        model_name = attrs['_name'] if '_name' in attrs else class_name_to_model_name(name)
+        is_new_model = not any(name == base.__name__ or model_name == base._name for base in bases)
+
+        base_class = bases[0]
+        if bases[0] not in OdooBaseModels:
+            base_class = base_class.__base__
+            if is_new_model:
+                raise TypeError(f'The new Model {name!r} must contain the Odoo model type (AbstractModel, Model, TransientModel) before other inherited models.')
```

Comment:
```suggestion
                raise TypeError(f'The new Model {name!r} must contain the Odoo model type {OdooBaseModels} before other inherited models.')
```

Reply:
`{OdooBaseModels}` is not very readable

Reply:
Then `{tuple(cls.__name__ for cls in OdooBaseModels)}` or similar.

Reply:
display: `('BaseModel', 'Model', 'TransientModel')` instead of ('AbstractModel', 'Model', 'TransientModel')